perf(neo4j): improve neo4j query performance by using node labels

pashashaik-mms opened the pull request:
**PR created and contributed by:** MediamarktSaturn Technology GmbH, Analytics-Services Team. Special thanks to @raudzis for the finding and idea proposed.

**PR Introduction:**
This PR introduces an optimization to the Neo4j querying process within our Datahub project. Previously, our Neo4j queries did not specify node labels during the match phase, which resulted in scanning all nodes in the database. This approach was inefficient, especially for large datasets. By integrating dynamic node labels into our match queries, we significantly improve query performance by leveraging Neo4j's ability to use indexes more effectively.

- **Node Label Integration:** Modified the Neo4j queries wherever applicable and now, the query explicitly targets nodes with the specified label, reducing the search space and improving performance.

- **Performance:** By applying node labels directly in our match clauses, the database engine can optimize node lookups using existing indexes, thus speeding up the query execution by reducing the number of nodes scanned.

- **Scalability:** These improvements make our database queries more scalable, handling larger datasets more efficiently.

- **Maintainability:** This change also enhances the clarity of our queries, making them more understandable at a glance, which benefits new contributors and maintainers alike.
-----------------------------------------------------------------------
## Checklist

- [X] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- ~[ ] Links to related issues (if applicable)~
- [X] Tests for the changes have been added/updated (if applicable)
- ~[ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.~
- ~[ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)~

pashashaik-mms commented:
This PR introduces an optimization to the Neo4j querying process within our Datahub project. Previously, our Neo4j queries did not specify node labels during the match phase, which resulted in scanning all nodes in the database. This approach was inefficient, especially for large datasets. By integrating dynamic node labels into our match queries, we significantly improve query performance by leveraging Neo4j's ability to use indexes more effectively.

**Node Label Integration:** Modified the Neo4j queries wherever applicable and now, the query explicitly targets nodes with the specified label, reducing the search space and improving performance.

**Performance:** By applying node labels directly in our match clauses, the database engine can optimize node lookups using existing indexes, thus speeding up the query execution by reducing the number of nodes scanned.
**Scalability:** These improvements make our database queries more scalable, handling larger datasets more efficiently.
**Maintainability:** This change also enhances the clarity of our queries, making them more understandable at a glance, which benefits new contributors and maintainers alike.

You can find the screenshot attached comparing the profile result:

**Profile WITHOUT NODELABEL query**
![image](https://github.com/datahub-project/datahub/assets/85875002/6aa0ca8b-4a78-44c5-8936-23a564dd0520)

**Profile WITH NODELABEL query**
![image](https://github.com/datahub-project/datahub/assets/85875002/d8093cbc-be88-4a5c-89f4-2c94f14c2166)

david-leifker reviewed (changes requested):
There are broken tests with `./gradlew :metadata-io:test`

pankajmahato-visa commented:
@pashashaik-mms 
 Run './gradlew :metadata-io:spotlessApply' to fix these violations.

pashashaik-mms commented:
@RyanHolstien @david-leifker Could you please approve it as I had fixed the changes. It was a formatting issue and hence the build was failing and so are the unit-tests. I fixed it now and would need your approval.

pashashaik-mms commented:
> Could you please approve it as I had fixed the changes. It was a formatting issue and hence the build was failing and so are the unit-tests. I fixed it now and would need your approval.

deepgarg-visa commented:
@pashashaik-mms , as mentions [here](https://github.com/datahub-project/datahub/actions/runs/9198346655/job/25300909269?pr=10415#step:6:3427)

the below error is occurring because in testcases [here](https://github.com/datahub-project/datahub/blob/master/metadata-io/src/test/java/com/linkedin/metadata/graph/GraphServiceTestBase.java#L1108), "sourceEntityFilter" is passed as null and because of that the variable "srcNodeLabel" in method "findRelatedEntities" is not setting having a default value of blank, which results in the below query which is not correct.

`MATCH (src: )-[r:DownstreamOf ]-(dest ) WHERE left(type(r), 2)<>'r_'  RETURN dest, type(r) SKIP $offset LIMIT $count"`

Please also handle the case where variable "sourceEntityFilter" in method "findRelatedEntities" can be null or empty

`metadata-io > nonsearch > com.linkedin.metadata.graph.dgraph.DgraphGraphServiceTest > testFindRelatedEntitiesDestinationType[11](dataset, [HasOwner], {or=[], direction=UNDIRECTED}, [RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset1,PROD), via=null), RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset2,PROD), via=null), RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset3,PROD), via=null), RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset4,PROD), via=null)]) STANDARD_ERROR
    org.neo4j.bolt.protocol.common.fsm.error.TransactionStateTransitionException: Invalid input ')': expected "%", "(" or an identifier (line 1, column 13 (offset: 12))
    "MATCH (src: )-[r:DownstreamOf ]-(dest ) WHERE left(type(r), 2)<>'r_'  RETURN dest, type(r) SKIP $offset LIMIT $count"`

pashashaik-mms commented:
> @pashashaik-mms , as mentions [here](https://github.com/datahub-project/datahub/actions/runs/9198346655/job/25300909269?pr=10415#step:6:3427)
> 
> the below error is occurring because in testcases [here](https://github.com/datahub-project/datahub/blob/master/metadata-io/src/test/java/com/linkedin/metadata/graph/GraphServiceTestBase.java#L1108), "sourceEntityFilter" is passed as null and because of that the variable "srcNodeLabel" in method "findRelatedEntities" is not setting having a default value of blank, which results in the below query which is not correct.
> 
> `MATCH (src: )-[r:DownstreamOf ]-(dest ) WHERE left(type(r), 2)<>'r_' RETURN dest, type(r) SKIP $offset LIMIT $count"`
> 
> Please also handle the case where variable "sourceEntityFilter" in method "findRelatedEntities" can be null or empty
> 
> `metadata-io > nonsearch > com.linkedin.metadata.graph.dgraph.DgraphGraphServiceTest > testFindRelatedEntitiesDestinationType[11](dataset, [HasOwner], {or=[], direction=UNDIRECTED}, [RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset1,PROD), via=null), RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset2,PROD), via=null), RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset3,PROD), via=null), RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset4,PROD), via=null)]) STANDARD_ERROR org.neo4j.bolt.protocol.common.fsm.error.TransactionStateTransitionException: Invalid input ')': expected "%", "(" or an identifier (line 1, column 13 (offset: 12)) "MATCH (src: )-[r:DownstreamOf ]-(dest ) WHERE left(type(r), 2)<>'r_' RETURN dest, type(r) SKIP $offset LIMIT $count"`

FIXED NOW

pashashaik-mms commented:
@deepgarg-visa fixed now. Could you please check and approve it. Its been waiting a while now.

deepgarg-visa commented:
@pashashaik-mms are all metadata-io testcase passed ?

deepgarg-visa commented on metadata-io/src/main/java/com/linkedin/metadata/graph/neo4j/Neo4jGraphService.java:
```diff
@@ -535,16 +529,40 @@ public RelatedEntitiesResult findRelatedEntities(
+        matchTemplate = "MATCH (src:%s %s)-[r%s %s]->(dest %s)%s";
+      }
+      // Build Statement strings
+      baseStatementString =
```
this is common in both if and else. It can be put outside

pashashaik-mms commented on metadata-io/src/main/java/com/linkedin/metadata/graph/neo4j/Neo4jGraphService.java:
```diff
@@ -648,18 +666,34 @@ public void removeEdgesFromNode(
 
     // build node label from entity type
     final String srcNodeLabel = urn.getEntityType();
+    String matchTemplate = "";
```
After a short vacation, I could see that you had already these changes. I made it in sync. What should be considered for this PR? Should we close it? or keep it for approval?

david-leifker commented:
Looks like the changes were added in a different PR, closing this one. Thanks!
